### PR TITLE
Increased the max supported dimensions via source gen to 30 dimensions

### PR DESCRIPTION
**Pull request by tekian**
…ions via source-gen to 30 dimensions

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/extensions/pull/4580)

### CODE REVIEW

File: test/Generators/Microsoft.Gen.Metrics/Unit/ParserTests.StrongTypes.cs [hunk 1]
```diff
@@ -590,4 +591,72 @@ public static partial class MetricClass
...
+
+    [Fact]
+    public async Task MaxDimensions()
+    {
+        StringBuilder sb = new StringBuilder();
+        int i = 1;
+        for (; i < 30; i++)
+        {
+            sb.AppendLine($"public class C{i} : C{i + 1} {{ public string dim{i} {{get;set;}}}}");
+        }
```
Comment (xakep139) [inline test/Generators/Microsoft.Gen.Metrics/Unit/ParserTests.StrongTypes.cs:603]:
```suggestion
        }

```

**Comment by dariusclay** [comment]
Does this code still have the quirk where it counts properties with nested classes as dimensions themselves? IIRC it wasn't resolved internally either. It seems counter-intuitive behavior, so we should probably document it somewhere or fix it.

File: src/Generators/Microsoft.Gen.Metrics/Parser.cs [hunk 2]
```diff
@@ -656,7 +656,7 @@ private void Diag(DiagnosticDescriptor desc, Location? location, params object?[
                 ex.NamedType.ToDisplayString());
         }
 
-        if (strongTypeAttributeParameters.StrongTypeConfigs.Count > MaxTagNames)
+        if (strongTypeAttributeParameters.DimensionHashSet.Count > MaxTagNames)
```
Comment (xakep139) [inline src/Generators/Microsoft.Gen.Metrics/Parser.cs:659]:
```suggestion
        if (strongTypeAttributeParameters.TagHashSet.Count > MaxTagNames)
```

File: test/Generators/Microsoft.Gen.Metrics/Unit/ParserTests.StrongTypes.cs [hunk 3]
```diff
@@ -590,4 +591,73 @@ public static partial class MetricClass
...
+                public string Dim12 { get; set; }
+                public string Dim13 { get; set; }
+                public string Dim14 { get; set; }
+                public string Dim15 { get; set; }
+                public string Dim16 { get; set; }
+                public string Dim17 { get; set; }
+                public string Dim18 { get; set; }
+                public string Dim19 { get; set; }
+                public string Dim20 { get; set; }
+                public MyClassB MyTransitiveProperty { get; set; }
```
Comment (dariusclay) [inline test/Generators/Microsoft.Gen.Metrics/Unit/ParserTests.StrongTypes.cs:641]: I believe this property here is also "counted" as a dimension/tag.

File: src/Generators/Microsoft.Gen.Metrics/StrongTypeAttributeParameters.cs [hunk 4]
```diff
@@ -11,7 +11,7 @@ internal sealed class StrongTypeAttributeParameters
     public string MetricNameFromAttribute = string.Empty;
     public HashSet<string> TagHashSet = [];
     public Dictionary<string, string> TagDescriptionDictionary = [];
-    public List<StrongTypeConfig> StrongTypeConfigs = [];
+    public List<StrongTypeConfig> DimensionHashSet = [];
```
Comment (xakep139) [inline src/Generators/Microsoft.Gen.Metrics/StrongTypeAttributeParameters.cs:14]: Please revert this change

File: src/Generators/Microsoft.Gen.Metrics/Parser.cs (hunk 2 above)
Comment (dariusclay) [reply src/Generators/Microsoft.Gen.Metrics/Parser.cs:659]: The terminology exposed is "Tag" rather than "Dimension", correct?

File: test/Generators/Microsoft.Gen.Metrics/Unit/ParserTests.StrongTypes.cs (hunk 3 above)
Comment (xakep139) [reply test/Generators/Microsoft.Gen.Metrics/Unit/ParserTests.StrongTypes.cs:641]: No it's not 🙂

**Review by xakep139 — APPROVED** [review]

File: src/Generators/Microsoft.Gen.Metrics/Parser.cs (hunk 2 above)
Comment (tekian) [reply src/Generators/Microsoft.Gen.Metrics/Parser.cs]: Is it OK now?

File: src/Generators/Microsoft.Gen.Metrics/StrongTypeAttributeParameters.cs (hunk 4 above)
Comment (tekian) [reply src/Generators/Microsoft.Gen.Metrics/StrongTypeAttributeParameters.cs]: Done.

**Review by dariusclay — APPROVED** [review]

**Review by geeknoid — APPROVED** [review]
